### ISSUE TITLE
Automate building templates

### ISSUE THREAD

**Issue by ericoporto** [adventuregamestudio/ags-template-source#27]
With agspak it should be possible to package the templates from command line. 

[agspak usage](
https://github.com/adventuregamestudio/ags/blob/master/Tools/agspak/main.cpp#L20-L23)
```
Usage: agspak <input-dir> <output-pak> [OPTIONS]
Options:
   -p <MB>        split game assets between partitions of this size max
   -r             recursive mode: include all subdirectories too
```

So passing `agspak -r BASS BASS.agt` from the template sources would for example pack BASS template.

There's a small detail that we need a release of the command line tools, we don't have a task for this yet. Alternatively, we build it from source and use it.

**Comment by ericoporto** [comment]
We could just fetch the AGS source code at a known alright commit/tag as .tar.gz, and use it to build and provide release from here.

- here would be easier to just have a directory for the 3.6.0 branch and a different one for the 3.5.1 branch, and release both to assets, which can then be fetched in AGS repository
- we can also use branches, but then the code that fetches the templates better be a script on the engine/editor codebase, since it will have to pickup the branch name from the GH API of assets release information to select the right asset to use.
  - There's though a command line tool from GH themselves that could help making things even easier

Alternatively, we could pull this repository of templates sources directly and pack the AGS templates in the AGS pipeline.

- we have branches here but only maintain the packaging in the AGS CI
- this is less code to maintain, the CI already have agspak source too
- in the future when we add building of templates source, for purposes of checking the code is correct, it's possible a broken commit is picked up in engine/Editor side of things, which is bad
- it's more things dependant on Cirrus-CI, which already does a lot

**Comment by morganwillcock** [comment]
> There's a small detail that we need a release of the command line tools, we don't have a task for this yet. Alternatively, we build it from source and use it.

Am I right in thinking that there is now a working script compiler + all the other tools that are needed to build a game and assembly the pieces using just command line utilities?

If so, maybe it is possible to:
* add the compiler and utility builds to the CI build on the main AGS repostory and add them to the release assets
* add compilation and game build to the CI build on the template source repository by downloading the compiler from an AGS release
* if template compilation does not cause an error then pack as a template and upload as a CI asset
* optionally create releases from the same assets

**Comment by ericoporto** [comment]
> Am I right in thinking that there is now a working script compiler + all the other tools that are needed to build a game and assembly the pieces using just command line utilities?

in theory we would need also either some tool or script to orchestrate this, so currently this is not easy. If we want to only use the script compiler to check the scripts and not necessarily assembly a working game, it would work more easily. 

Alternatively, AGS Editor with `/compile` flag does run using GitHub Actions, I have been using this in my modules. The empty template is the only one that won't work in this way since it has no rooms.

The template which is only a "zip-like" using ags format can be easily achieved today with the mentioned command line at the top though.

The difficulty with compilation as error checking is we can only update the templates one release after, so they need to be always one version behind - I am ok with this, just mentioning the main difference.

**Comment by morganwillcock** [comment]
> in theory we would need also either some tool or script to orchestrate this

Would a Makefile do or is it more complicated than that?

**Comment by ericoporto** [comment]
uhm, if you manually set the modules and stuff, maybe a Makefile could do. Below is an example bat script @ivan-mogilko once sent me:

```batch
setlocal enabledelayedexpansion
set BIN=%1

mkdir temp
set WORK=temp

copy Game.agf %WORK%
copy agsdefns.sh %WORK%
copy *.ash %WORK%
copy *.asc %WORK%
copy *.crm %WORK%

cd %WORK%

rem Game Auto headers
%BIN%\agf2autoash.exe Game.agf _AutoGenerated.ash
%BIN%\agf2glvar.exe Game.agf _GlobalVariables.ash _GlobalVariables.asc
%BIN%\agf2dlgasc.exe Game.agf DialogScript.asc

rem Room Auto headers
for /r %%f in (*.crm) do %BIN%\crm2ash.exe "%%f" "%%f.ash"

rem Prepare modules
set AUTO_HEADERS=agsdefns.sh:_AutoGenerated.ash:_GlobalVariables.ash
set MODULES=_GlobalVariables Utils CameraTricks DragDrop ObjectPool KeyboardMovement GlobalScript
set MODULE_HEADERS=

(for %%m in (%MODULES%) do (
    SET ASC=%%m.asc
    SET OUT=%%m.o
    %BIN%\agscc.exe !ASC! !OUT! -H %AUTO_HEADERS%!MODULE_HEADERS!
    SET MODULE_HEADERS=!MODULE_HEADERS!:%%m.ash
))
```

I also have an unfinished tool that generates a ninja script of the same - by looking into `Game.agf` directly.

**Comment by ivan-mogilko** [comment]
@morganwillcock yes, i think it should be possible to compile the game fully using existing tools, except, like @ericoporto said, you will have to provide a list of script files, and maybe some other files, yourself. Either hardcode in script, or write one universal script that reads lists of these source files from a text file.

**Comment by morganwillcock** [comment]
> To clarify a little, if you do not want to compile anything, but only package the template, you'd only require tool "agspak".
Template file is same package format as game data. (Same of room templates)

It is mainly just to perform some check that indicates a script error or warning (I think that in theory the Tumbleweed template will be emitting warnings?). So just the script compiling would probably do as a starting point and also highlight the gaps the tooling.

Also, I think there is probably a build issue with the utilities in the main AGS repository because the last time I tried I wasn't able to compile them, and getting those built and available from somewhere in some form is probably a good idea if they are meant to be used.

**Comment by ericoporto** [comment]
@morganwillcock what sort of warnings should the Tumbleweed template emit? I verified today and could not see any.

**Comment by morganwillcock** [comment]
The old style extender declarations, but since the fix was just applied that should no longer be a problem (and I'm not sure whether the compiler actual implements output for warnings or not).

I did have a look at creating a Makefile that matches the batch file posted above. Do you know if the ordering of the modules in the list is critical to it working? I think the benefit of the new compiler was that it removed the need for explicit ordering, is this command line frontend done using that compiler?

**Comment by ericoporto** [comment]
module order independent was this PR here: https://github.com/adventuregamestudio/ags/pull/1408 (it's in ags4 not master)

The compiler frontend is in master, so currently old compiler, so order matters for script module headers. If you need any warning behavior I can add - I thought I had added, so I need to check how this works in the old compiler.

as I mentioned it's possible also to just use ags editor itself with the compile flag if needed, example here: https://github.com/ericoporto/DungeonHands

**Comment by ivan-mogilko** [comment]
"old style extender declarations" was probably with the new ags4 compiler too.

**Comment by ivan-mogilko** [comment]
Concerned of the lack of progress here, and because there's a problem #40, I had an alternate idea: support templates packed as simple zip files:
https://github.com/adventuregamestudio/ags/issues/1998

This will remove a need to build and use ags tools for template packaging, we'd only need a zip.

What do you think about this?

**Comment by ericoporto** [comment]
I don't understand, the issue is not the packaging, if we skip building for checking, the problem is the different ags branches and how to release from here and pull in ags repository (see my second comment)

**Comment by ivan-mogilko** [comment]
The packaging is also an issue, as you need to have a ready packaging tool. If you use agspak, then you have a dependency on ags repository. If you use zip, then you don't have a dependency.

**Comment by ericoporto** [comment]
What are the feelings towards adding a command line option in the Editor to export templates? Then it would work here by using a Windows CI until other things are solved.

That would be easy to add to the Editor - I think.

**Comment by ivan-mogilko** [comment]
> What are the feelings towards adding a command line option in the Editor to export templates? Then it would work here by using a Windows CI until other things are solved.

What prevents from using agspak tool in a similar way?

**Comment by ericoporto** [comment]
> There's something strange about this whole issue, simply uploading agspak somewhere where CI can get it would let start automated building of templates few years ago

Well that is true, it's just the that if the tool doesn't make a release from the main repo I have nowhere to fetch it here. Perhaps there could be a repository for generic file storage?

**Comment by ericoporto** [comment]
OK, I made a way to add a template export option in the editor here: https://github.com/ericoporto/ags/commit/3f1909b1095878feb21d17f20ae885bc029058ca

I also alternatively added template export in my [agstoolbox tool](https://github.com/ericoporto/agstoolbox/blob/main/src/agstoolbox/core/ags/ags_template.py).

**Comment by ericoporto** [comment]
I started work in [agstoolbox to support also building templates with the editor](https://github.com/ericoporto/agstoolbox/commit/a1fdd8ee5a7c63f926908a0e05bbb6d4dcdfcf02) since the `/template` command now made into releases. I need to test a bit more and if it's all working, I should upgrade the agstoolbox in the ci, and then the automated builds can really kick-in as it will be done with the editor instead.

**Comment by ericoporto** [comment]
OK, now need to merge the master changes to ags4 branch and fix the conflicts plus do the version upgrade in tumbleweed template.
